### PR TITLE
"getting started" tutorial

### PR DESCRIPTION
**Pull request by joelgrus**
I don't know where tutorials should live in the long term. In the meantime I stuck all of the relevant files under `tutorials/getting_started`. This PR does the following:

(1) first attempt at a "Getting Started" tutorial that involves training a POS tagger on a small subset of the Brown Corpus
(2) experiment definition for the getting started task (I just guessed at all the hyperparameters, I'm sure there's better ones)
(3) train/dev/test datasets under `tutorial/getting_started/data`
(4) created a `BrownCorpusDatasetReader` and wrote tests for it
(5) plumbed `evaluate` into the command line (that step got skipped somehow)
(6) added `non_padded_namespaces` as config option to `train` (there are so many tags in the Brown corpus that each test set seems to have novel ones so I need an UNKNOWN token in the tags namespace too :P)
(7) added "accuracy" and "top 3 accuracy" as metrics to `simple_tagger.py`
(8) fixed an edge case in top k categorical accuracy that caused a crash if you make fewer than k predictions (this happened in tests, for example)

ugh, that's a lot for one PR, but writing the tutorial required all of those changes

### CODE REVIEW

File: allennlp/commands/train.py [hunk 1]
```diff
@@ -129,7 +133,7 @@ def train_model(param_dict: Dict[str, Any]):
     train_data = dataset_reader.read(train_data_path)
 
     # TODO(Mark): work out how this is going to be built with different options.
-    vocab = Vocabulary.from_dataset(train_data)
+    vocab = Vocabulary.from_dataset(train_data, non_padded_namespaces=non_padded_namespaces)
```
Comment (matt-gardner) [inline allennlp/commands/train.py]: Yeah, this is what Mark meant in his TODO comment.  We need to add a `from_params` option here, and add a `vocab` key to the top-level config.  It's a little complicated, though, because of different options for instantiating this object (e.g., we need something like `from_dataset_and_params`, but done in a sane way that doesn't have too many ways of creating the object), so I wouldn't worry about fixing it in this PR.  I'll add a github issue for this.

File: allennlp/commands/train.py [hunk 2]
```diff
@@ -108,7 +109,10 @@ def train_model(param_dict: Dict[str, Any]):
     params = Params(replace_none(param_dict))
     prepare_environment(params)
 
-    log_dir = params.get("serialization_prefix", None)  # pylint: disable=no-member
+    trainer_params = params.pop("trainer")
+    log_dir = trainer_params.get("serialization_prefix")
```
Comment (matt-gardner) [inline allennlp/commands/train.py]: Hmm, looks like I broke this with a recent PR without realizing it.  I was wondering why the logging output of my recent BiDAF training run wasn't saved like I expected.  Is there some way we can test this against current example experiment configs?  It would have been harder for me to miss the fact that I broke this.  I'm not sure what test would have caught this particular error, though...

File: allennlp/data/dataset_readers/brown_corpus.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@DatasetReader.register("brown_corpus")
+class BrownCorpusDatasetReader(DatasetReader):
+    """
+    Reads instances from the Brown Corpus where each line is in the following format:
+
+    WORD/TAG WORD/TAG WORD/TAG ... WORD/TAG
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/brown_corpus.py]: What if `WORD` or `TAG` has a `/`?  I think the logic below needs to do a `pair.rsplit("/", 1)`, and this needs a comment that `/` is not allowed as a character in `TAG`.

File: allennlp/data/dataset_readers/brown_corpus.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+from allennlp.data.instance import Instance
+from allennlp.data.token_indexers.token_indexer import TokenIndexer
+from allennlp.data.fields import TextField, TagField
+from allennlp.data.token_indexers import SingleIdTokenIndexer
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@DatasetReader.register("brown_corpus")
+class BrownCorpusDatasetReader(DatasetReader):
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/brown_corpus.py]: Can we just use the `SequenceTaggingDatasetReader` here, instead of creating a new one?  If the only reason for making a new class is that the delimiter is different, we could just change the delimiter in the `SequenceTaggingReader`, or make it configurable.  The only reason the delimiter is `###` in the other reader is to disallow fewer tags / make collisions less likely.  But if the delimiter is configurable, this isn't really an issue.

File: allennlp/data/dataset_readers/brown_corpus.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        """
+        Parameters
+        ----------
+        token_indexers: ``Dict[Params]``, optional
+        """
+        token_indexers = {}
+        token_indexer_params = params.pop('token_indexers', Params({}))
+        for name, indexer_params in token_indexer_params.items():
+            token_indexers[name] = TokenIndexer.from_params(indexer_params)
+        # The default parameters are contained within the class,
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/brown_corpus.py]: Looks like all of our readers have this logic duplicated - can we put it somewhere central and re-use it?  Either on `TokenIndexer` as a method like `indexer_dict_from_params`, or something, or in `DatasetReader` as a helper function to process these arguments.  Also happy for this to be raised as a new issue if you don't want to deal with it in this PR.

File: tests/data/dataset_readers/brown_corpus_dataset_reader_test.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# pylint: disable=no-self-use,invalid-name
+from allennlp.data.dataset_readers import BrownCorpusDatasetReader
+from allennlp.common.testing import AllenNlpTestCase
+
+
+class TestBrownCorpusDatasetReader(AllenNlpTestCase):
+    def test_read_from_file(self):
+        reader = BrownCorpusDatasetReader()
+        dataset = reader.read('tests/fixtures/data/brown_corpus.txt')
```
Comment (matt-gardner) [inline tests/data/dataset_readers/brown_corpus_dataset_reader_test.py]: If we re-use the sequence tagging reader instead of creating a new one, I'd still keep this fixture and this test, just make sure we can read the brown corpus with the right delimiter setting using the existing reader.

**Comment by schmmd** [comment]
@joelgrus I think that a top-level folder `tutorials` makes sense.  Can you move Mark's notebooks into this folder?  I think it's odd that they are at `allennlp/notebooks` presently.

File: tutorials/getting_started/getting_started.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+(If your machine has GPUs, use `allenlp-gpu` instead.)
+
+This will download the latest `allennlp` image to your machine
+(unless you already have it),
+start a Docker container, and launch an interactive shell.
+(It also exposes port 8000, which is where the demo server runs,
+ and shuts down the container when you exit the interactive shell.)
+
+## Installing Not Using Docker
```
Comment (schmmd) [inline tutorials/getting_started/getting_started.md]: I would delete this section, and instead refer people to the README if they want an alternate installation.

File: tutorials/getting_started/getting_started.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+```js
+  "trainer": {
+    "num_epochs": 20,
+    "serialization_prefix": "/tmp/tutorials/getting_started",
+    "cuda_device": -1
+  }
+```
+
+The `num_epochs` parameter just specifies how many passes through the dataset we want to train on.
```
Comment (schmmd) [inline tutorials/getting_started/getting_started.md]: I would explain what an epoch is a bit clearer here.  What is the tradeoff between having more/fewer epochs other than time?

File: tutorials/getting_started/getting_started.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+accuracy: 0.50, accuracy_top3: 0.64, loss: 2.27 ||: 100%|##########| 50/50 [00:00<00:00, 53.72it/s]
+2017-08-15 11:38:19,609 - INFO - allennlp.training.trainer - Training accuracy : 0.477715    Validation accuracy : 0.501286
+2017-08-15 11:38:19,610 - INFO - allennlp.training.trainer - Training accuracy3 : 0.631720    Validation accuracy3 : 0.642796
+2017-08-15 11:38:19,610 - INFO - allennlp.training.trainer - Training loss : 2.187194    Validation loss : 2.265297
+2017-08-15 11:38:19,617 - INFO - allennlp.training.trainer - Best validation performance so far. Copying weights to /tmp/tutorials/getting_started/best.th'.
+```
+
+Here `accuracy` measures how often our model predicted the "correct" part of speech tag as most probable,
+while `accuracy3` measures how often the correct tag was one of the _three_ most probable.
+(`loss` measures [cross entropy](https://en.wikipedia.org/wiki/Cross_entropy), which you probably don't care about
```
Comment (schmmd) [inline tutorials/getting_started/getting_started.md]: maybe state explicitly what it's a measure of.  "You probably don't care about the implementation details, but cross entropy measures whether the model is improving.  You should see that it's mostly decreasing during training."

File: tutorials/getting_started/getting_started.md [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+accuracy: 0.71, loss: 1.38, accuracy3: 0.83 ||: 100%|##########| 50/50 [00:01<00:00, 47.64it/s]
+2017-08-15 13:22:28,214 - INFO - allennlp.training.trainer - Training accuracy : 0.849197    Validation accuracy : 0.714408
+2017-08-15 13:22:28,214 - INFO - allennlp.training.trainer - Training loss : 0.577224    Validation loss : 1.378879
+2017-08-15 13:22:28,214 - INFO - allennlp.training.trainer - Training accuracy3 : 0.924702    Validation accuracy3 : 0.831475
+2017-08-15 13:22:28,222 - INFO - allennlp.training.trainer - Best validation performance so far. Copying weights to /tmp/tutorials/getting_started/best.th'.
+```
+
+This means that 71% of the time our model predicted the correct tag on the validation dataset,
+and 83% of the time the correct tag was in the model's "top 3".
+(Not ground-breaking performance, but this is a pretty simple model, and
```
Comment (schmmd) [inline tutorials/getting_started/getting_started.md]: Is performance low because it's a simple model, our training data is small, or the number of epochs we're using is small?

File: tutorials/getting_started/getting_started.md [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+if you look at the data there's a lot of different tags!)
+
+Now that the model is trained, there should be a bunch of files in the serialization directory. The `vocabulary` directory
+contains the model's vocabularies and (in particular) their encodings as integers. The various
```
Comment (schmmd) [inline tutorials/getting_started/getting_started.md]: This sentence confuses me--isn't a vocabulary exactly a mapping of words to integers?  "The `vocabulary` directory contains the model's vocabularies, which are encodings of the words as integers."

**Comment by schmmd** [comment]
Actually @joelgrus the tutorials should probably live on the website branch, not `master/tutorials`.  Let's catch up in person briefly.

**Comment by matt-gardner** [comment]
+1 for having tutorials / notebooks in the website repo instead of in here.

File: allennlp/data/dataset_readers/brown_corpus.py (hunk 4 above)
Comment (joelgrus) [inline allennlp/data/dataset_readers/brown_corpus.py]: done

File: allennlp/data/dataset_readers/brown_corpus.py (hunk 3 above)
Comment (joelgrus) [inline allennlp/data/dataset_readers/brown_corpus.py]: this file is gone now

File: allennlp/data/dataset_readers/brown_corpus.py (hunk 5 above)
Comment (joelgrus) [inline allennlp/data/dataset_readers/brown_corpus.py]: let me do this in my other PR, since I'm already tweaking this code

File: tests/data/dataset_readers/brown_corpus_dataset_reader_test.py (hunk 6 above)
Comment (joelgrus) [inline tests/data/dataset_readers/brown_corpus_dataset_reader_test.py]: 👍

File: allennlp/commands/train.py (hunk 2 above)
Comment (joelgrus) [inline allennlp/commands/train.py]: I mean, as long as `serialization_prefix` is allowed to be `None`, then it's not really an error when it ends up being `None`. Are we sure we want to allow that? What's the point of training without a serialization directory?

File: tutorials/getting_started/getting_started.md (hunk 7 above)
Comment (joelgrus) [inline tutorials/getting_started/getting_started.md]: For the "Getting Started" tutorial I'd like to keep the friction as low as possible. I'd hate for people not to try out AllenNLP because they don't want to install Docker.

File: tutorials/getting_started/getting_started.md (hunk 8 above)
Comment (joelgrus) [inline tutorials/getting_started/getting_started.md]: I don't want to say too much, I think our users know what an epoch is, but I reworded it a bit

File: tutorials/getting_started/getting_started.md (hunk 9 above)
Comment (joelgrus) [inline tutorials/getting_started/getting_started.md]: reworded

File: tutorials/getting_started/getting_started.md (hunk 11 above)
Comment (joelgrus) [inline tutorials/getting_started/getting_started.md]: reworded

File: tutorials/getting_started/getting_started.md (hunk 10 above)
Comment (joelgrus) [inline tutorials/getting_started/getting_started.md]: the first two. it seems like performance on the holdout set pretty much levels out by 20 epochs, so more training wouldn't help.

**Comment by joelgrus** [comment]
disagree: the tutorial has both code (or at least config files) and data, and I don't think those should live on the website branch. And also I feel like someone who clones `master` should get the tutorial with it. Agree that a HTML version of the tutorial should live on the website branch tho.

File: tests/data/dataset_readers/brown_corpus_dataset_reader_test.py (hunk 6 above)
Comment (matt-gardner) [inline tests/data/dataset_readers/brown_corpus_dataset_reader_test.py]: Can you move this into `sequence_tagging_dataset_reader_test.py`?

File: allennlp/data/dataset_readers/sequence_tagging.py [hunk 12]
```diff
@@ -23,18 +23,27 @@ class SequenceTaggingDatasetReader(DatasetReader):
...
 
-    and converts it into a ``Dataset`` suitable for sequence tagging.
+    and converts it into a ``Dataset`` suitable for sequence tagging. You can also specify
+    alternative delimiters in the constructor.
 
     Parameters
     ----------
+    word_tag_delimiter: ``str``, optional (default=``"###"``)
+        The text that separates each WORD from its TAG.
+    word_word_delimiter: ``str``, optional (default=``"\t"``)
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/sequence_tagging.py]: Can we have the default delimiter just be any whitespace?  Like, have this default to `None`, which will just call `line.split()`, instead of passing in a delimiter?  I'd also probably just call this `token_delimiter` - seems simpler.

File: allennlp/data/dataset_readers/sequence_tagging.py [hunk 13]
```diff
@@ -23,18 +23,27 @@ class SequenceTaggingDatasetReader(DatasetReader):
...
         We use this to define the input representation for the text.  See :class:`TokenIndexer`.
         Note that the `output` tags will always correspond to single token IDs based on how they
         are pre-tokenised in the data file.
     """
     def __init__(self,
+                 word_tag_delimiter: str = "###",
+                 word_word_delimiter: str = "\t",
                  token_indexers: Dict[str, TokenIndexer] = None) -> None:
         self._token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer()}
+        self.word_tag_delimiter = word_tag_delimiter
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/sequence_tagging.py]: Preference for private variables here, unless you really want to make this part of the public API.

File: allennlp/data/dataset_readers/sequence_tagging.py [hunk 14]
```diff
@@ -23,18 +23,27 @@ class SequenceTaggingDatasetReader(DatasetReader):
 
     WORD###TAG [TAB] WORD###TAG [TAB] ..... \n
 
-    and converts it into a ``Dataset`` suitable for sequence tagging.
+    and converts it into a ``Dataset`` suitable for sequence tagging. You can also specify
+    alternative delimiters in the constructor.
 
     Parameters
     ----------
+    word_tag_delimiter: ``str``, optional (default=``"###"``)
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/sequence_tagging.py]: `from_params` below needs to be updated to read these parameters.

File: allennlp/commands/train.py (hunk 2 above)
Comment (matt-gardner) [inline allennlp/commands/train.py]: Yeah, we should just require this to be not `None`.  I agree.

File: tutorials/getting_started/getting_started.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+## Installing using Docker
+
+The easiest way to get started is using Docker. Assuming you have Docker installed, just run
+
+```bash
+docker run -p 8000:8000 -it --rm allennlp/allennlp-cpu
+```
+
+(If your machine has GPUs, use `allenlp-gpu` instead.)
```
Comment (matt-gardner) [inline tutorials/getting_started/getting_started.md]: I'd just remove the parentheses here and on line 18.

File: tutorials/getting_started/getting_started.md [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+If you visit `http://localhost:8000` in your browser, you can play around with the same demo
+that runs on the AllenNLP website.
+
+TODO(joelgrus): screenshot
+
+### Training a Model
+
+In this tutorial we'll train a simple part-of-speech tagger using AllenNLP.
+The model is defined in [allennlp/models/simple_tagger.py](https://github.com/allenai/allennlp/blob/master/allennlp/models/simple_tagger.py).
+It consists of a word embedding layer followed by a LSTM.
```
Comment (matt-gardner) [inline tutorials/getting_started/getting_started.md]: `s/a LSTM/an LSTM/`

File: tutorials/getting_started/getting_started.md [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+accuracy: 0.50, accuracy_top3: 0.64, loss: 2.27 ||: 100%|##########| 50/50 [00:00<00:00, 53.72it/s]
+2017-08-15 11:38:19,609 - INFO - allennlp.training.trainer - Training accuracy : 0.477715    Validation accuracy : 0.501286
+2017-08-15 11:38:19,610 - INFO - allennlp.training.trainer - Training accuracy3 : 0.631720    Validation accuracy3 : 0.642796
+2017-08-15 11:38:19,610 - INFO - allennlp.training.trainer - Training loss : 2.187194    Validation loss : 2.265297
+2017-08-15 11:38:19,617 - INFO - allennlp.training.trainer - Best validation performance so far. Copying weights to /tmp/tutorials/getting_started/best.th'.
+```
+
+Here `accuracy` measures how often our model predicted the "correct" part of speech tag as most probable,
+while `accuracy3` measures how often the correct tag was one of the _three_ most probable.
+(`loss` measures [cross entropy](https://en.wikipedia.org/wiki/Cross_entropy)
```
Comment (matt-gardner) [inline tutorials/getting_started/getting_started.md]: I'd remove parentheses here, too.

File: tutorials/getting_started/getting_started.md [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+accuracy: 0.71, loss: 1.38, accuracy3: 0.83 ||: 100%|##########| 50/50 [00:01<00:00, 47.64it/s]
+2017-08-15 13:22:28,214 - INFO - allennlp.training.trainer - Training accuracy : 0.849197    Validation accuracy : 0.714408
+2017-08-15 13:22:28,214 - INFO - allennlp.training.trainer - Training loss : 0.577224    Validation loss : 1.378879
+2017-08-15 13:22:28,214 - INFO - allennlp.training.trainer - Training accuracy3 : 0.924702    Validation accuracy3 : 0.831475
+2017-08-15 13:22:28,222 - INFO - allennlp.training.trainer - Best validation performance so far. Copying weights to /tmp/tutorials/getting_started/best.th'.
+```
+
+This means that 71% of the time our model predicted the correct tag on the validation dataset,
+and 83% of the time the correct tag was in the model's "top 3".
+(Not ground-breaking performance, but this is a pretty simple model, and
```
Comment (matt-gardner) [inline tutorials/getting_started/getting_started.md]: Remove parens.

File: tutorials/getting_started/getting_started.md [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+if you look at the data there's a lot of different tags!)
+
+Now that the model is trained, there should be a bunch of files in the serialization directory. The `vocabulary` directory
+contains the model's vocabularies, each of which is a (distinct) encoding of strings as integers.
+In our case, we'll have one for `tokens` (i.e. words) and another for `tags`. The various
+`training_state_epoch_XX.th` files contain the state of the trainer after each epoch (`.th` is the suffix for serialized torch tensors),
+so that you could resume training where you left off, if you wanted to.
+Similarly, the `model_state_epoch_XX.th` files contain the model weights after each epoch.
+Finally `best.th` contains the *best* weights (that is, those from the epoch with the smallest `loss`).
```
Comment (matt-gardner) [inline tutorials/getting_started/getting_started.md]: Would be good to mention that best is defined by _validation_ loss.

File: tutorials/getting_started/getting_started.md [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+Once you've trained a model, you likely want to evaluate it on another dataset.
+In [tutorials/getting_started/data/cp.test](https://github.com/allenai/allennlp/blob/master/tutorials/getting_started/data/cp.test)
+we have the contents of Brown Corpus `cp01` (the first file in the "Romance" section).  As our configuration file specifies where
+the model was serialized to, it also tells us enough to evaluate the model:
+
+```
+allennlp/run evaluate --config_file tutorials/getting_started/simple_tagger.json --evaluation_data_file tutorials/getting_started/data/cp.test
+```
+
+After again spitting out a bunch of parameter values, it will evaluate the trained model on the evaluation data file:
```
Comment (matt-gardner) [inline tutorials/getting_started/getting_started.md]: Might make sense to disable parameter and initialization logging in the `evaluate` command.  I did this in `common/testing/test_case.py`, if you want to disable the most verbose logging statements with a copy/paste.

File: allennlp/commands/train.py (hunk 2 above)
Comment (joelgrus) [inline allennlp/commands/train.py]: ok, I added a check that throws a `ConfigurationError` if it's not specified
